### PR TITLE
Support non-str Hashables in DataArray

### LINKED ISSUE #8546: DataArray types must be strings
### What happened?

While trying to understand the way dimensions are typed in `namedarray` I was confused by the requirement that dimension names must be strings:

```python
>>> da = xr.DataArray(data=[1,2,3], dims=[7])
TypeError: dimension 7 is not a string
```

However, the type hints suggest that dimension names can more generally be `Hashable`. According to @headtr1ck's [post](https://github.com/pydata/xarray/pull/8294#issuecomment-1852714645), the desired behavior is that `Hashable` dimension names should be allowed, so this `TypeError` may be considered a bug.

### What did you expect to happen?

_No response_

### Minimal Complete Verifiable Example

_No response_

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 5.15.0-89-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C.UTF-8
LANG: C.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.2
libnetcdf: 4.9.2

xarray: 2023.11.0
pandas: 1.5.3
numpy: 1.26.2
scipy: 1.11.4
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: 2.16.1
cftime: 1.6.3
nc_time_axis: None
iris: None
bottleneck: None
dask: 2023.11.0
distributed: 2023.11.0
matplotlib: 3.8.2
cartopy: 0.22.0
seaborn: 0.13.0
numbagg: None
fsspec: 2023.10.0
cupy: None
pint: None
sparse: 0.14.0
flox: 0.8.3
numpy_groupies: 0.10.2
setuptools: 68.2.2
pip: 23.3.1
conda: 23.10.0
pytest: 7.4.3
mypy: 1.7.1
IPython: 8.18.0
sphinx: None
[/opt/conda/lib/python3.11/site-packages/_distutils_hack/__init__.py:33](https://vscode-remote+dev-002dcontainer-002b7b22686f737450617468223a222f686f6d652f6d617265732f7265706f732f6174747269627574696f6e222c226c6f63616c446f636b6572223a66616c73652c22636f6e66696746696c65223a7b22246d6964223a312c22667350617468223a222f686f6d652f6d617265732f7265706f732f6174747269627574696f6e2f2e646576636f6e7461696e65722f646576636f6e7461696e65722e6a736f6e222c2265787465726e616c223a2266696c653a2f2f2f686f6d652f6d617265732f7265706f732f6174747269627574696f6e2f2e646576636f6e7461696e65722f646576636f6e7461696e65722e6a736f6e222c2270617468223a222f686f6d652f6d617265732f7265706f732f6174747269627574696f6e2f2e646576636f6e7461696e65722f646576636f6e7461696e65722e6a736f6e222c22736368656d65223a2266696c65227d7d.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/_distutils_hack/__init__.py:33): UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

</details>

### PR DESCRIPTION
**Pull request by headtr1ck**
- [x] Closes #8546
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Probably we should add a whole bunch of tests for this. For now only testing the constructor.

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -140,7 +131,9 @@ def _check_coords_dims(shape, coords, dim):
 
 
 def _infer_coords_and_dims(
-    shape, coords, dims
+    shape: tuple[int, ...],
+    coords: Sequence[Sequence[Any] | pd.Index | DataArray] | Mapping[Any, Any] | None,
+    dims: str | Iterable[Hashable] | None,
```
Comment (max-sixty) [inline xarray/core/dataarray.py:136]: We have a `Dims` type we could use?

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -140,7 +131,9 @@ def _check_coords_dims(shape, coords, dim):
 
 
 def _infer_coords_and_dims(
-    shape, coords, dims
+    shape: tuple[int, ...],
+    coords: Sequence[Sequence[Any] | pd.Index | DataArray] | Mapping[Any, Any] | None,
```
Comment (max-sixty) [inline xarray/core/dataarray.py:135]: Very minor, but unless we're explicitly stating that these can contain `Any`, I would vote to keep them small:

```suggestion
    coords: Sequence[Sequence | pd.Index | DataArray] | Mapping | None,
```

**Comment by TomNicholas** [comment]
> Probably we should add a whole bunch of tests for this. For now only testing the constructor.

I think I've said this before but eventually I would like to see us use property-based testing and test using a strategy that generates arbitrary strings / hashables. We could build off of https://github.com/pydata/xarray/pull/6903 and generalise the `names` strategy that was introduced in https://github.com/pydata/xarray/pull/8404, for example.

**Comment by max-sixty** [comment]
> Probably we should add a whole bunch of tests for this.

We could add them onto a canonical test object like https://github.com/pydata/xarray/pull/8533 (maybe there's a param for "object with lots of quirks"), and then if we can spread that object throughout the test suite, we get every method tested on the full range of quirks

(I see @TomNicholas just replied with a suggestion to use hypothesis — agree — I think that would provide more depth (more quirks!), though less wide coverage, given we're unlikely to write every test using hypothesis)

File: xarray/core/dataarray.py (hunk 1 above)
Comment (headtr1ck) [reply xarray/core/dataarray.py:136]: Unfortunately that contains ellipsis.

**Comment by headtr1ck** [comment]
Should we explicitly disallow using `None` as dimension/variable names?
It will inadvertently lead to problems later on.

File: xarray/tests/test_hashable.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from typing import TYPE_CHECKING, Union
+
+import pytest
+
+from xarray import DataArray, Dataset, Variable
+
+if TYPE_CHECKING:
+    from xarray.core.types import TypeAlias
+
+    DimT: TypeAlias = Union[int, tuple, "DEnum", "CustmomHashable"]
```
Comment (max-sixty) [inline xarray/tests/test_hashable.py:13]:
```suggestion
    DimT: TypeAlias = Union[int, tuple, "DEnum", "CustomHashable"]
```

:)

**Pull request merged**